### PR TITLE
spec version 0.4.0 update: shard_block_root -> crosslink_data_root; switch from validator.slashed_epoch -> validator.slashed approach; a couple assert -> doAssert; note fix for underflow-prone spec changes in checkAttestation; GENESIS_SLOT change from 2^63 to 2^32; refactor Proposer and ProposalSlashing structure; add signed_root; remove duplicate slashValidator

### PR DESCRIPTION
**Pull request by tersec**
I saw https://github.com/status-im/nim-beacon-chain/pull/158 but it's apparently having issues in CI?

So these conflict in a couple places. Not sure which the best order would be to apply them, but didn't want to wait to see how the CI issues would resolve.

### CODE REVIEW

File: beacon_chain/spec/beaconstate.nim [hunk 1]
```diff
@@ -143,13 +126,14 @@ func slash_validator*(state: var BeaconState, index: ValidatorIndex) =
     whistleblower_reward = get_effective_balance(state, index) div
       WHISTLEBLOWER_REWARD_QUOTIENT
 
+  ## TODO here and elsewhere, if reduce_balance can't reduce balance by full
```
Comment (arnetheduck) [inline beacon_chain/spec/beaconstate.nim]: my take would be that the validator should be gone by the time this is an issue - ie there's a minimum balance where they get ejected well before the whistleblower reward becomes an issue, this way..

File: beacon_chain/ssz.nim [hunk 2]
```diff
@@ -290,6 +290,18 @@ func hash_tree_root*[T: object|tuple](x: T): array[32, byte] =
     for field in x.fields:
       h.update hash_tree_root(field)
 
+# https://github.com/ethereum/eth2.0-specs/blob/0.4.0/specs/simple-serialize.md#signed-roots
+func signed_root*[T: object](x: T, field_name: string): array[32, byte] =
+  # As of 0.4.0, this is the only valid field_name.
+  # TODO write tests for this (check vs hash_tree_root)
```
Comment (arnetheduck) [inline beacon_chain/ssz.nim]: maybe the (potentially static) assert in here should be that we actually find the field.. intuitively, this should be possible to do at compile time, but that can be left for later..

File: beacon_chain/spec/beaconstate.nim [hunk 3]
```diff
@@ -380,25 +357,26 @@ func update_validator_registry*(state: var BeaconState) =
...
     state: BeaconState, attestation: Attestation, flags: UpdateFlags): bool =
   ## Check that an attestation follows the rules of being included in the state
   ## at the current slot. When acting as a proposer, the same rules need to
   ## be followed!
 
-  # Can't underflow, because GENESIS_SLOT > MIN_ATTESTATION_INCLUSION_DELAY
-  doAssert GENESIS_SLOT > MIN_ATTESTATION_INCLUSION_DELAY
+  if not (attestation.data.slot >= GENESIS_SLOT):
+    warn("Attestation predates genesis slot",
+      attestation_slot = humaneSlotNum(attestation.data.slot),
```
[the remark this answers is not in the document]
Comment (tersec) [reply beacon_chain/spec/beaconstate.nim]: Fixed.

File: beacon_chain/ssz.nim (hunk 2 above)
Comment (tersec) [reply beacon_chain/ssz.nim]: One current approach would be to check against specific types (it's used with, I think, 3 object types total). That's pretty unappealing for obvious reasons, though.

Instead, I just did a quick and kind of ugly flag variable, so make this in theory general purpose (and implicitly do at least some kind of testing that it's working as intended, in advance of actual tests).

**Pull request merged**